dnf/transaction_sr.py

```diff
@@ -0,0 +1,167 @@
+new transaction JSON).
+
+MINOR version denotes extending the format without breaking backwards
+compatibility (old dnf can work with new transaction JSON), but breaks forwards
+compatibility (new dnf won't work with old transaction JSON).
+"""
+
+
+class TransactionError(dnf.exceptions.Error):
+    def __init__(self, msg):
+        super(TransactionError, self).__init__(msg)
+
+
+class TransactionFileError(dnf.exceptions.Error):
+    def __init__(self, filename, errors):
+        """
+        :param filename: The name of the transaction file being replayed
+        :param errors: a list of error classes or a string with an error description
+        """
+        if isinstance(errors, (list, tuple)):
+            if len(errors) > 1:
+                msg = _('Errors in "{filename}":').format(filename=filename)
+                for error in errors:
+                    msg += "\n  " + str(error)
+
+                super(TransactionFileError, self).__init__(msg)
+                return
+
+            else:
+                errors = str(errors[0])
+
+        msg = _('Error in "{filename}": {error}').format(filename=filename, error=errors)
+        super(TransactionFileError, self).__init__(msg)
+
+
+class IncompatibleTransactionVersionError(TransactionFileError):
+    def __init__(self, filename, msg):
+        super(IncompatibleTransactionVersionError, self).__init__(filename, msg)
+
+
+def check_version(version, filename):
+    major, minor = version.split('.')
+    if major != VERSION_MAJOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible major version {major}, supported major version is {major_supp}.')
+                .format(major=major, major_supp=VERSION_MAJOR)
+        )
+
+    if minor < VERSION_MINOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible minor version {minor}, '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+    }
+    rpms = {}
+    groups = {}
+    environments = {}
+
+    for tsi in transaction.packages():
+        if tsi.is_package():
+            na = "%s.%s" % (tsi.name, tsi.arch)
+
+            pkg_data = rpms.setdefault(na, {})
+
+            # for installonly packages the package data needs to be a list of actions
+            if isinstance(pkg_data, list):
+                pkg_data.append({})
+                pkg_data = pkg_data[-1]
+
+            if tsi._item.isForwardAction():
+                if "to" in pkg_data or "action" in pkg_data and pkg_data["action"] == "Removed":
+                    # installonly pkg, convert to a list of actions
+                    rpms[na] = [pkg_data, {}]
+                    pkg_data = rpms[na][-1]
+
+                pkg_data["to"] = tsi.evr
```

Comment:
Is tsi.evr always with epoch or not?

Reply:
It appears it's left out if epoch is 0.

Reply:
No problem, it means that it uses the same behaviour like rpm package class, because libsolv is handling EVR and EPOCH by this way.